File: storage/note_store.py
"""
笔记存储层
支持 Markdown 文件存储 + SQLite 索引
"""

import os
import sqlite3
import json
from datetime import datetime
from pathlib import Path
from config import NOTES_DIR, DB_PATH, ALLOWED_TAGS, DEFAULT_TAG, SUMMARY_MAX_CHARS


class NoteStore:
    def __init__(self):
        self.notes_dir = Path(NOTES_DIR)
        self.notes_dir.mkdir(parents=True, exist_ok=True)
        self._init_db()

    def _init_db(self):
        """初始化 SQLite"""
        conn = sqlite3.connect(DB_PATH)
        c = conn.cursor()
        c.execute("""
            CREATE TABLE IF NOT EXISTS notes (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                content TEXT NOT NULL,
                summary TEXT,
                tags TEXT,
                source TEXT DEFAULT 'qq',
                created_at TEXT NOT NULL,
                updated_at TEXT NOT NULL
            )
        """)
        # 全文搜索索引
        c.execute("CREATE INDEX IF NOT EXISTS idx_content ON notes(content)")
        c.execute("CREATE INDEX IF NOT EXISTS idx_tags ON notes(tags)")
        c.execute("CREATE INDEX IF NOT EXISTS idx_created ON notes(created_at)")
        conn.commit()
        conn.close()

    def add_note(self, content: str, tags: list = None, summary: str = None):
        """添加笔记"""
        tags = tags or [DEFAULT_TAG]
        tags_str = ",".join(tags)

        # 摘要截断
        if summary and len(summary) > SUMMARY_MAX_CHARS:
            summary = summary[:SUMMARY_MAX_CHARS] + "..."

        now = datetime.now().isoformat()
        conn = sqlite3.connect(DB_PATH)
        c = conn.cursor()
        c.execute("""
            INSERT INTO notes (content, summary, tags, created_at, updated_at)
            VALUES (?, ?, ?, ?, ?)
        """, (content, summary, tags_str, now, now))
        note_id = c.lastrowid
        conn.commit()
        conn.close()

        # 同时存一份 Markdown 方便直接查看
        self._save_as_markdown(note_id, content, tags, summary, now)

        return note_id

    def _save_as_markdown(self, note_id: int, content: str, tags: list, summary: str, created_at: str):
        """存一份 Markdown 文件"""
        date = created_at[:10]
        filename = self.notes_dir / f"{date}_{note_id}.md"
        tags_str = " ".join([f"`{t}`" for t in tags])
        with open(filename, 'w', encoding='utf-8') as f:
            f.write(f"# 笔记 #{note_id}\n\n")
            f.write(f"**标签**：{tags_str}\n\n")
            f.write(f"**摘要**：{summary or '无'}\n\n")
            f.write(f"**时间**：{created_at}\n\n")
            f.write("---\n\n")
            f.write(content)

    def search(self, keyword: str, limit: int = 10) -> list:
        """搜索笔记"""
        conn = sqlite3.connect(DB_PATH)
        c = conn.cursor()
        c.execute("""
            SELECT id, content, summary, tags, created_at
            FROM notes
            WHERE content LIKE ?
            ORDER BY created_at DESC
            LIMIT ?
        """, (f"%{keyword}%", limit))
        results = c.fetchall()
        conn.close()
        return results

    def get_recent(self, limit: int = 10) -> list:
        """获取最近笔记"""
        conn = sqlite3.connect(DB_PATH)
        c = conn.cursor()
        c.execute("""
            SELECT id, content, summary, tags, created_at
            FROM notes
            ORDER BY created_at DESC
            LIMIT ?
        """, (limit,))
        results = c.fetchall()
        conn.close()
        return results

    def get_by_tag(self, tag: str, limit: int = 10) -> list:
        """按标签查询"""
        conn = sqlite3.connect(DB_PATH)
        c = conn.cursor()
        c.execute("""
            SELECT id, content, summary, tags, created_at
            FROM notes
            WHERE tags LIKE ?
            ORDER BY created_at DESC
            LIMIT ?
        """, (f"%{tag}%", limit))
        results = c.fetchall()
        conn.close()
        return results


if __name__ == "__main__":
    store = NoteStore()
    # 测试添加
    note_id = store.add_note(
        content="这是一个测试笔记，关于 Vue3 的响应式原理",
        tags=["技术", "前端", "Vue"],
        summary="Vue3 响应式原理"
    )
    print(f"添加成功，ID: {note_id}")

    # 测试搜索
    results = store.search("Vue3")
    print(f"搜索结果: {len(results)} 条")

File: src/storage/noteStore.ts
import Database from 'better-sqlite3';
import * as fs from 'fs';
import * as path from 'path';
import { NOTES_DIR, DB_PATH, ALLOWED_TAGS, DEFAULT_TAG, SUMMARY_MAX_CHARS } from '../config';

export interface Note {
  id: number;
  content: string;
  summary: string | null;
  tags: string;
  source: string;
  created_at: string;
  updated_at: string;
}

export interface NoteResult {
  id: number;
  content: string;
  summary: string | null;
  tags: string;
  created: string;
}

export class NoteStore {
  private db: Database.Database;

  constructor() {
    if (!fs.existsSync(NOTES_DIR)) {
      fs.mkdirSync(NOTES_DIR, { recursive: true });
    }
    this.db = new Database(DB_PATH);
    this.initDb();
  }

  private initDb(): void {
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS notes (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        content TEXT NOT NULL,
        summary TEXT,
        tags TEXT,
        source TEXT DEFAULT 'qq',
        created_at TEXT NOT NULL,
        updated_at TEXT NOT NULL
      )
    `);
    this.db.exec(`CREATE INDEX IF NOT EXISTS idx_content ON notes(content)`);
    this.db.exec(`CREATE INDEX IF NOT EXISTS idx_tags ON notes(tags)`);
    this.db.exec(`CREATE INDEX IF NOT EXISTS idx_created ON notes(created_at)`);
  }

  addNote(content: string, tags: string[], summary: string | null = null): number {
    const tagsStr = tags.join(',');
    let finalSummary = summary;
    if (finalSummary && finalSummary.length > SUMMARY_MAX_CHARS) {
      finalSummary = finalSummary.substring(0, SUMMARY_MAX_CHARS) + '...';
    }

    const now = new Date().toISOString();
    const stmt = this.db.prepare(`
      INSERT INTO notes (content, summary, tags, source, created_at, updated_at)
      VALUES (?, ?, ?, 'qq', ?, ?)
    `);
    const result = stmt.run(content, finalSummary, tagsStr, now, now);
    const noteId = result.lastInsertRowid as number;

    this.saveAsMarkdown(noteId, content, tags, finalSummary, now);
    return noteId;
  }

  private saveAsMarkdown(noteId: number, content: string, tags: string[], summary: string | null, createdAt: string): void {
    const date = createdAt.substring(0, 10);
    const filename = path.join(NOTES_DIR, `${date}_${noteId}.md`);
    const tagsStr = tags.map(t => `\`${t}\``).join(' ');

    const md = `# 笔记 #${noteId}\n\n` +
      `**标签**：${tagsStr}\n\n` +
      `**摘要**：${summary || '无'}\n\n` +
      `**时间**：${createdAt}\n\n` +
      `---\n\n` +
      `${content}`;

    fs.writeFileSync(filename, md, 'utf-8');
  }

  search(keyword: string, limit: number = 10): NoteResult[] {
    const stmt = this.db.prepare(`
      SELECT id, content, summary, tags, created_at
      FROM notes
      WHERE content LIKE ?
      ORDER BY created_at DESC
      LIMIT ?
    `);
    const rows = stmt.all(`%${keyword}%`, limit) as Note[];
    return rows.map(r => ({
      id: r.id,
      content: r.content.length > 50 ? r.content.substring(0, 50) + '...' : r.content,
      summary: r.summary,
      tags: r.tags,
      created: r.created_at.substring(0, 10),
    }));
  }

  getRecent(limit: number = 10): NoteResult[] {
    const stmt = this.db.prepare(`
      SELECT id, content, summary, tags, created_at
      FROM notes
      ORDER BY created_at DESC
      LIMIT ?
    `);
    const rows = stmt.all(limit) as Note[];
    return rows.map(r => ({
      id: r.id,
      content: r.content.length > 40 ? r.content.substring(0, 40) + '...' : r.content,
      summary: r.summary,
      tags: r.tags,
      created: r.created_at.substring(5, 10),
    }));
  }

  getByTag(tag: string, limit: number = 10): NoteResult[] {
    const stmt = this.db.prepare(`
      SELECT id, content, summary, tags, created_at
      FROM notes
      WHERE tags LIKE ?
      ORDER BY created_at DESC
      LIMIT ?
    `);
    const rows = stmt.all(`%${tag}%`, limit) as Note[];
    return rows.map(r => ({
      id: r.id,
      content: r.content.length > 40 ? r.content.substring(0, 40) + '...' : r.content,
      summary: r.summary,
      tags: r.tags,
      created: r.created_at.substring(5, 10),
    }));
  }
}
